database/api_test.go

```diff
@@ -112,15 +115,225 @@ func (matcher *queryMatcher) Matches(x interface{}) bool {
 		return false
 	}
 	sql, _, err := query.SQL(qb.NoLimit, 0)
-	assert1.NoError(matcher.t, err)
-	assert1.Equal(matcher.t, matcher.sql, sql)
+	_require.NoError(matcher.t, err)
+	_require.Equal(matcher.t, matcher.sql, sql)
 	return true
 }
 
 func (matcher *queryMatcher) String() string {
 	return "*qb.SelectQuery"
 }
 
+func Test_SelectWithTotal(t *testing.T) {
+
+	var (
+		ctrl   = gomock.NewController(t)
+		api    = NewMockAPI(ctrl)
+		target TestRecordCollection
+		limit  = 0
+		offset = 0
+	)
+
+	query := qb.Select(MetaTestRecord.ID).From(MetaTestRecord)
+
+	t.Run("limit equal zero", func(t *testing.T) {
+		require := _require.New(t)
+		api.EXPECT().Count(MetaTestRecord, query).Return(int32(2), nil)
+		_, total, err := SelectWithTotal(api, MetaTestRecord, target, query, limit, offset)
+		require.NoError(err)
+		require.Equal(2, total)
+	})
+
+	t.Run("total equal zero", func(t *testing.T) {
+		require := _require.New(t)
+		limit = 1 // change limit
+		api.EXPECT().Count(MetaTestRecord, query).Return(int32(0), nil)
+		_, total, err := SelectWithTotal(api, MetaTestRecord, target, query, limit, offset)
+		require.NoError(err)
+		require.Equal(0, total)
+	})
+
+	t.Run("total equal 1", func(t *testing.T) {
+		require := _require.New(t)
+		api.EXPECT().Count(MetaTestRecord, query).Return(int32(1), nil)
+		api.EXPECT().Select(&target, query, record.NewListOptions(limit, offset)).Return(nil)
+		_, total, err := SelectWithTotal(api, MetaTestRecord, target, query, limit, offset)
+		require.NoError(err)
+		require.Equal(1, total)
+	})
+
+	t.Run("error case: Count error", func(t *testing.T) {
+		require := _require.New(t)
+		expected := generator.String(20)
+		api.EXPECT().Count(MetaTestRecord, query).Return(int32(0), errors.New(expected))
+		_, total, err := SelectWithTotal(api, MetaTestRecord, target, query, limit, offset)
+		require.Equal(0, total)
+		require.EqualError(err, expected)
+	})
+
+	t.Run("error case: Select error", func(t *testing.T) {
+		require := _require.New(t)
+		expected := generator.String(20)
+		api.EXPECT().Count(MetaTestRecord, query).Return(int32(1), nil)
+		api.EXPECT().Select(&target, query, record.NewListOptions(limit, offset)).Return(errors.New(expected))
+		_, total, err := SelectWithTotal(api, MetaTestRecord, target, query, limit, offset)
+		require.EqualError(err, expected)
+		require.Equal(0, total)
```

Comment:
when not testing error values we want to use `assert` so the execution continues. We use `require` when we want to stop testing immediately (e.g. when an error occurs and further execution will npe).